torch/autograd/gradcheck.py

```diff
@@ -1531,14 +1534,15 @@ def randn_like(x):
         tupled_grad_outputs = _as_tuple(grad_outputs)
 
     num_outputs = len(tupled_grad_outputs)
+    diff_input_args_indices = set(i for i, x in enumerate(tupled_inputs) if is_tensor_like(x) and x.requires_grad)
 
     def new_func(*args):
-        input_args = args[:-num_outputs]
-        grad_outputs = args[-num_outputs:]
+        # Make sure inputs requires_grad=True here because forward AD detaches
+        input_args = tuple(x.requires_grad_() if i in diff_input_args_indices else x for i, x in enumerate(args[:-num_outputs]))
         outputs = _differentiable_outputs(func(*input_args))
-        input_args = tuple(x for x in input_args
-                           if is_tensor_like(x) and x.requires_grad)
-        grad_inputs = torch.autograd.grad(outputs, input_args, grad_outputs, create_graph=True,
+        grad_outputs = tuple(x.requires_grad_() for x in args[-num_outputs:])
+        diff_input_args = tuple(x for i, x in enumerate(input_args) if i in diff_input_args_indices)
```

Comment:
You just fixed the requires_grad fields, so no need to change this?

Reply:
yeah there should be no need, but this one I prefer because we can avoid repeating checking is_tensor_like and requires grad again. diff_input_args also better reflects what the tuple contains compared to input_args